File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.py
class Solution:
    def LIS(self,previdx,idx,nums,dp):
        if idx>=len(nums):
            return 0 
        if dp[previdx][idx]!=-1:
            return dp[previdx][idx]
        if nums[idx]>nums[previdx]:
            a=1+self.LIS(idx,idx+1,nums,dp)
            b=self.LIS(previdx,idx+1,nums,dp)
            dp[previdx][idx]=max(a,b)
            return dp[previdx][idx] 
        dp[previdx][idx]=self.LIS(previdx,idx+1,nums,dp)
        return dp[previdx][idx]

    def lengthOfLIS(self, nums: List[int]) -> int:
        # at each pt you can it can be start or increaseing sequence 
        # computer for each value LIS then you can use that reference again 
        ans = 0
        n = len(nums)
        dp = [ [ -1 for _ in range(n) ] for _ in range(n)]
        for i in range(len(nums)-1,-1,-1):
            res = 1+self.LIS(i,i+1,nums,dp)
            ans=max(res,ans)
        return ans

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.cpp
class Solution {
public:
    int maxLIS(vector<int>& nums,int previdx,int curridx,vector<vector<int>>& dp){
        if((curridx-1)==nums.size()) {
            return 0;
        }
        if(dp[previdx][curridx]!=-1) return dp[previdx][curridx];
        // if(previdx!=-1 && dp[previdx]!=-1) return dp[previdx];

        if((previdx-1)==-1 ||  nums[curridx-1]>nums[previdx-1]){
            return dp[previdx][curridx]=max(
                1+maxLIS(nums,curridx,curridx+1,dp),
                maxLIS(nums,previdx,curridx+1,dp)
            );
        } 
        
        return dp[previdx][curridx]=maxLIS(nums,previdx,curridx+1,dp);
    }

    int lengthOfLIS(vector<int>& nums) {
        int n = nums.size();
        vector<vector<int>> dp(n+2,vector<int>(n+2,-1));
        return maxLIS(nums,0,1,dp);
    }
};